[clang] Lower non-builtin sincos[f|l] calls to llvm.sincos.* when -fno-math-errno is set

MacDue opened the pull request:
This will allow vectorizing these calls (after a few more patches). This should not change the codegen for targets that enable the use of AA during the codegen (in `TargetSubtargetInfo::useAA()`). This includes targets such as AArch64. This notably does not include x86 but can be worked around by passing `-mllvm -combiner-global-alias-analysis=true` to clang.

Follow up to #114086.

MacDue committed:
[clang] Lower non-builtin sincos[f|l] calls to llvm.sincos.* when -fno-math-errno is set

This will allow vectorizing these calls (after a few more patches). This
should not change the codegen for targets that enable the use of AA
during the codegen (in `TargetSubtargetInfo::useAA()`). This includes
targets such as AArch64. This notably does not include x86 but can be
worked around by passing `-mllvm -combiner-global-alias-analysis=true`
to clang.

arsenm commented on clang/test/CodeGen/AArch64/sincos.c:
```diff
@@ -1,5 +1,19 @@
+// RUN: %clang_cc1 -triple=aarch64-gnu-linux -emit-llvm -O1 %s -o - -DUSE_C_DECL | FileCheck --check-prefix=NO-MATH-ERRNO %s
+// RUN: %clang_cc1 -triple=aarch64-gnu-linux -emit-llvm -fmath-errno %s -o - -DUSE_C_DECL | FileCheck --check-prefix=MATH-ERRNO %s
+
+#if defined(USE_BUILTIN)
```
Can you just duplicate the test functions into builtin and non-builtin forms? This is much harder to follow

MacDue replied:
Sure, done :+1:

MacDue committed:
Tweak tests

sdesmalen-arm reviewed (commented):
@RKSimon do you know if there is a reason CodeGen AA is not enabled for x86?

I'd otherwise be happy to approve this patch, lowering to the intrinsic seems like the right thing to do.

RKSimon commented:
> @RKSimon do you know if there is a reason CodeGen AA is not enabled for x86?
> 
> I'd otherwise be happy to approve this patch, lowering to the intrinsic seems like the right thing to do.

I don't remember any reason for alias analysis to not be enabled on x86 - @phoebewang @topperc do you know?

topperc commented:
> > @RKSimon do you know if there is a reason CodeGen AA is not enabled for x86?
> > I'd otherwise be happy to approve this patch, lowering to the intrinsic seems like the right thing to do.
> 
> I don't remember any reason for alias analysis to not be enabled on x86 - @phoebewang @topperc do you know?

I don't remember either.

phoebewang commented:
> > > @RKSimon do you know if there is a reason CodeGen AA is not enabled for x86?
> > > I'd otherwise be happy to approve this patch, lowering to the intrinsic seems like the right thing to do.
> > 
> > 
> > I don't remember any reason for alias analysis to not be enabled on x86 - @phoebewang @topperc do you know?
> 
> I don't remember either.

Me neither. @clin111 do you happen to know it?

MacDue commented:
> > > > @RKSimon do you know if there is a reason CodeGen AA is not enabled for x86?
> > > > I'd otherwise be happy to approve this patch, lowering to the intrinsic seems like the right thing to do.
> > > 
> > > 
> > > I don't remember any reason for alias analysis to not be enabled on x86 - @phoebewang @topperc do you know?
> > 
> > 
> > I don't remember either.
> 
> Me neither. @clin111 do you happen to know it?

I've created a patch enabling it here: https://github.com/llvm/llvm-project/pull/123787 (maybe you'll spot why it's disabled there :sweat_smile:). I've updated the generated tests, and gone over the handwritten ones more carefully, and updated them myself.

The handwritten tests that changed were:
```
llvm/test/CodeGen/X86/cfguard-x86-vectorcall.ll
llvm/test/CodeGen/X86/fixup-bw-inst.ll
llvm/test/CodeGen/X86/misched-matrix.ll
llvm/test/CodeGen/X86/regalloc-advanced-split-cost.ll
llvm/test/CodeGen/X86/vectorcall.ll
llvm/test/CodeGen/X86/win32-eh.ll
``` 

Someone more experienced with x86(_64) should double check the changes make sense. I am fairly confident the underlying transforms should be correct though as this feature is enabled for other targets (at least AArch64, RISC-V, off the top of my head).

clin111 commented:
Looks like it was never enabled --- or tested on X86 --- since UseAA was introduced in https://reviews.llvm.org/D67266. Eventually the dead option was removed with c266776. Would advise some caution with regards to register pressure and compile time.

MacDue commented:
Given it looks like #123787 won't be accepted (due to compile time regressions), I think we should restrict this lowering to AArch64 (and other supported targets) until there's a solution for targets without AA in codegen too. Does that seem reasonable?

MacDue committed:
Add hook to limit intrinsic use

MacDue commented:
I've pushed a change to restrict this to AArch64 (other targets could enable it too) given no reply above, or progress on https://github.com/llvm/llvm-project/pull/123787.

arsenm reviewed (changes requested):
I'm a hard no on making this a target dependent option. Just enable it unconditionally and take the regression

MacDue committed:
Revert "Add hook to limit intrinsic use"

This reverts commit aafed97efb71225893965378d2125085924a2293.

MacDue commented:
> I'm a hard no on making this a target dependent option. Just enable it unconditionally and take the regression

Fine, I don't mind the regression as it does not affect targets I work on, I just wanted to make the changes needed for AArch64 without knowingly slightly regressing other targets.

MacDue committed:
Add math-libcalls test
